source/transactions/transactions.rst

```diff
@@ -744,24 +753,59 @@ format:
 
     { ok : 1, writeConcernError: {code: <Number>, errmsg : "..."} }
 
-Mongos Pinning for Sharded Transactions
----------------------------------------
+Sharded Transactions
+--------------------
+
+MongoDB 4.2 (maxWireVersion 8) introduces support for sharded transactions.
+Sharded transactions support all of the same features as single replica set
+transaction but introduce two new driver concepts: mongos pinning and the
+``recoveryToken`` field.
 
-The server supports sharded transactions starting in MongoDB 4.2 (
-maxWireVersion 8). The server requires that drivers MUST send all commands for
-a single transaction to the same mongos.
+Mongos Pinning
+~~~~~~~~~~~~~~
+
+The server requires that drivers MUST send all commands for
+a single transaction to the same mongos (excluding retries of commitTransaction
+and abortTransaction) therefore this spec introduces the concept of pinning a
+ClientSession to a mongos.
 
 After the driver selects a mongos for the first command within a transaction,
 the driver MUST pin the ClientSession to the selected mongos. Drivers MUST
-send all commands that are part of the same transaction, including
-commitTransaction and abortTransaction and any retries thereof, to the
-same mongos.
+send all subsequent commands that are part of the same transaction (excluding
+retries of commitTransaction and abortTransaction) to the same mongos.
+
+When to unpin
+^^^^^^^^^^^^^
+
+A pinned ClientSession MUST be unpinned after running the initial
+commitTransaction or abortTransaction attempt regardless if that attempt
+succeeded or failed. When the commitTransaction or abortTransaction attempt
+fails on the pinned mongos, the session MUST be unpinned and any
+subsequent (automatic or explicit) retry attempt MUST perform normal mongos
+server selection. When a commitTransaction or abortTransaction
+attempt succeeds on the pinned mongos, the session MUST also be unpinned.
```

Comment:
Why do we recommend unpinning after sending commitTransaction once?

It seems better to remain pinned, because only the original mongos can actually commit the transaction - sending commitTransaction to a recovery mongos can *only* be used to *recover* the decision.

Reply:
If there's a retryable error we have to unpin to run a retry attempt. If the commit succeeds then an explicit retry can recover the previous outcome on any mongos using the recoveryToken. So optimistically unpinning the session after a commit regardless of the outcome seemed like a reasonable simplification.

However, I've now changed the spec to mandate unpinning after any transient error within a transaction so this "simplification" really no longer applies. I think I can remove this and instead say that a session must remain pinned after a successful commit (just like any other successful operation within a transaction). This would also avoid the [SERVER-39349](https://jira.mongodb.org/browse/SERVER-39349) problem that makes some of our spec tests run slower than they should.

Reply:
I removed the unpin after successful commit/abort rule and updated the tests.